File: multi_thread.py
'''
Assignment: Determining the number of "cores" are on a machine
Submitted by: Md Ishtiaq Ahmed
Objective: Run a program using multiple threads to determine how many cores are in a computer. 
Here, I created a function to fill up an array repeatedly and call the function using multiple threads.
If the system is using the cores equitably, run times should be until there are more than N threads (where N is the reported number of cores). 

'''

import threading 
import time  
import statistics  
import os

cpu_count = os.cpu_count() # Get the core count of the computer
print(f'Number of CPUs: {cpu_count}\n' )

# Function to fill the array
# Pre condition: None
# Post Condition: Filling up an array repeatedly
def fill_array():
    array_size = 10000  # Size of the array

    # Fill the array 100 times
    for _ in range(100):
        
        new_array = [i for i in range(array_size)] 
    # end for
# end fill_array

# Function to run the threads
# Pre condition: A function to fill up an array multiple times
# Post Condition: fill_array function will be executed using multiple threads and return the execution time       

def thread_count(num_threads_max):
    iteration = []  # List to store execution times of each iteration
    print(f'Running with {num_threads_max} thread:')

    # Run 5 iterations
    for i in range(1, 6):
        start_time = time.time()  # Get the start time of the iteration
        threads = []  # List to store the thread objects

        # Create and start multiple threads
        for _ in range(num_threads_max):
            # Create a thread targeting the fill_array function
            thread = threading.Thread(target=fill_array)
            threads.append(thread)  # Append the thread to the list
            thread.start()  # Start the thread 
        #end for

        # Wait for all threads to finish
        for thread in threads:
            thread.join() 
        # end for

        end_time = time.time()  # End time of the iteration

        # Calculate the execution time of the iteration in nanoseconds
        run_time = (end_time - start_time) * 1e9
        iteration.append(run_time)  # Append the execution time to the list

   
        print(f'Iteration {i} running time: {run_time} nanoseconds')
    # end for

    # Calculate and print the average and standard deviation
    standard_deviation = statistics.stdev(iteration) # Standard Deviation
    avg = statistics.mean(iteration) # Average 
    print(f'Average Running time using {num_threads_max} thread:  {avg} nanoseconds')
    print(f'Standard Deviation using {num_threads_max} thread: {standard_deviation} \n')
# end thread_count


# Run the thread_count function 
num_threads = int(input("Enter the maximum number of threads: ")) # get the max number of thread from user input
# Execute the thread_count function using 1 thread to num_thread
for i in range(1, num_threads+1):
    thread_count(i) # call the thread_count function 
#end for

File: multi_thread.java
/*
Assignment: Determining the number of "cores" are on a machine
Submitted by: Md Ishtiaq Ahmed
Objective: Run a program using multiple threads to determine how many cores are in a computer. 
Here, I created a function to fill up an array repeatedly and call the function using multiple threads.
If the system is using the cores equitably, run times should be until there are more than N threads (where N is the reported number of cores). 
*/
// Multithread code reference from https://stackoverflow.com/questions/5238397/populating-array-from-different-threads
// Multithread code reference from https://www.geeksforgeeks.org/multithreading-in-java/

import java.util.Arrays;
import java.util.Scanner;

//Pre condition: None
//Post condition: Execute the fill_array method using multiple thread for multiple iteration. Return average and standard deviation
// of runnning time for each thread 
public class multi_thread {
    public static void main(String[] args) {
        int cpuCount = Runtime.getRuntime().availableProcessors(); // Get the core count of the computer
        System.out.println("Number of CPUs: " + cpuCount);
        System.out.println();

        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the maximum number of threads: ");
        final int num_threads_max = scanner.nextInt(); // Maximum number of threads to use
        scanner.close();
        final int numIterations = 5; // Number of iterations per thread
        final int arraySize = 100000; // Size of the array

        // Iterate over each thread configuration
        for (int numThreads = 1; numThreads <= num_threads_max; numThreads++) {
            System.out.println("Running with " + numThreads + " thread:");
            long[] executionTimes = new long[numIterations]; // Array to store execution times for each iteration

            // Run multiple iterations
            for (int i = 0; i < numIterations; i++) {
                long startTime = System.nanoTime(); // Start time of the iteration
                fill_array(arraySize, numThreads); // Run the iteration
                long endTime = System.nanoTime(); // End time of the iteration
                long iterationTime = endTime - startTime; // Calculate iteration time
                executionTimes[i] = iterationTime; // Store iteration time in the array
                System.out.println("Iteration " + (i + 1) + " running time: " + iterationTime + " nanoseconds"); 
            } // end for

            // Calculate and print average and standard deviation
            double average = get_average(executionTimes); // Calculate average execution time
            double standard_deviation = get_std_dev(executionTimes); // Calculate standard deviation
            System.out.println("Average Running time using " + numThreads + " thread: " + average + " nanoseconds");
            System.out.println(
                    "Standard deviation using " + numThreads + " thread: " + standard_deviation + " nanoseconds"); 
 
            System.out.println();
        } // end for
    } // end main

    // Method to fill the array
    // Pre condition: None
    // Post Condition: Filling up an array repeatedly using multiple threads
    public static void fill_array(int arraySize, int numThreads) {
        Thread[] threads = new Thread[numThreads]; // Create an array of threads
       
        for (int i = 0; i < numThreads; i++) {
            // final int threadIndex = i;
            threads[i] = new Thread(() -> {
                int[] new_array = new int[arraySize];
                for (int j = 0; j < 10000; j++) // Outer loop for repeated filling
                {
                    for (int k = 0; k < arraySize; k++) // Inner loop to fill the array
                    {
                        new_array[k] = k; // Assigning values to array elements
                    }
                }
            });
            threads[i].start(); // Start each thread
        } // end for

        // Wait for all threads to finish
        for (Thread thread : threads) {
            try {
                thread.join(); // Wait for each thread to finish
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        } // end for
    } // end fill_array

    // Method to calculate average
    public static double get_average(long[] numbers) {
        long sum = 0;
        for (long num : numbers) {
            sum += num; // Sum up all the numbers
        }
        return (double) sum / numbers.length; // return the average
    }

    // Method to calculate standard deviation
    public static double get_std_dev(long[] numbers) {
        double mean = get_average(numbers); // Calculate the mean
        double squared_diff_sum = 0; // Variable to store sum of squared differences
        for (long num : numbers) { // Loop to calculate squared differences
            squared_diff_sum += (num - mean) * (num - mean); // Sum of squared differences
        }
        // Calculate the variance
        double variance = squared_diff_sum / numbers.length;
        return Math.sqrt(variance); // Return the standard deviation
    }
}